staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,458 @@
+	// to conflict with the subnet owning the conflicts cache. If subnet X is
+	// the owner of the conflicts cache and a subnet Y is in rivals this means
+	// that while a queue worker was working on Y it found a conflict with X and
+	// at that time X had the VNI and CIDR values in ownerData.
+	rivals []k8stypes.NamespacedName
+}
+
+// Validator performs validation for newly-created or updated subnets, and
+// writes in their status the outcome of the validation. Validation consists of
+// two checks:
+//
+// 		(1) CIDRs for subnets with the same VNI are disjoint.
+// 		(2) all subnets with the same VNI are within the same K8s namespace.
+//
+// If a subnet S1 does not pass validation because of a conflict with another
+// subnet S2, upon deletion of S2 S1 is validated again.
+// Validator uses an informer on Subnets to be notified of creation or updates,
+// but does a live list against the API server to retrieve the conflicting
+// subnets when validating a subnet, to avoid race conditions caused by multiple
+// validators running at the same time.
+type Validator struct {
+	netIfc         kosclientv1a1.NetworkV1alpha1Interface
+	subnetInformer k8scache.SharedInformer
+	subnetLister   netlistv1a1.SubnetLister
+	queue          k8sworkqueue.RateLimitingInterface
+	workers        int
+
+	// conflicts associates a subnet namespaced name with its conflictsCache.
+	// Always access while holding conflictsMutex.
+	conflicts      map[k8stypes.NamespacedName]*conflictsCache
+	conflictsMutex sync.Mutex
+
+	// staleRVs associates a subnet X's namespaced name for which there was a
+	// successful status update to X's resource version prior to the update.
+	// When a worker begins processing a subnet X, it checks whether X's
+	// resource version matches the resource version in staleRVs[X]. If that's
+	// the case X is stale, i.e. it does not reflect the latest update yet,
+	// hence processing is immediately stopped.
+	// Only access while holding staleRVsMutex.
+	staleRVs      map[k8stypes.NamespacedName]string
+	staleRVsMutex sync.Mutex
+}
+
+func NewValidator(netIfc kosclientv1a1.NetworkV1alpha1Interface,
+	subnetInformer k8scache.SharedInformer,
+	subnetLister netlistv1a1.SubnetLister,
+	queue k8sworkqueue.RateLimitingInterface,
+	workers int) *Validator {
+
+	return &Validator{
+		netIfc:         netIfc,
+		subnetInformer: subnetInformer,
+		subnetLister:   subnetLister,
+		queue:          queue,
+		workers:        workers,
+		conflicts:      make(map[k8stypes.NamespacedName]*conflictsCache),
+		staleRVs:       make(map[k8stypes.NamespacedName]string),
+	}
+}
+
+// Run starts the validator and blocks until stop is closed. This entails
+// starting its Informer and the worker goroutines.
+func (v *Validator) Run(stop <-chan struct{}) error {
+	defer k8sutilruntime.HandleCrash()
+	defer v.queue.ShutDown()
+
+	v.subnetInformer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
+		AddFunc:    v.OnSubnetCreate,
```

Comment:
If we just rename these methods to `OnAdd`, `OnUpdate`, and `OnDelete` then we can simply `v.subnetInformer.AddEventHandler(v)`.

Reply:
Oh. Nice! Done in 8ac9883.